Device Tree clock information

carlescufi opened the pull request:
First split out from #16469
Second in #16958

carlescufi commented:
cc @Sizurka

carlescufi commented:
Missing files to be updated by @pizi-nordic :

```
boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.dts
boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts
dts/arm/silabs/efr32mg.dtsi
dts/arm/silabs/efr32fg1p.dtsi
```
`system clock definitions for the ARC nsim boards`

carlescufi commented:
@galak replying to [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/16469#pullrequestreview-252305939) of yours:

>I'm trying to understand the usage here between the cpu clock-frequency, a system clock frequency and having full clock support for a given SoC.

From my limited understanding, that I am sure @pizi-nordic and @Sizurka can correct or expand, the CPU clock frequency is the actual clock frequency at which the CPU operates. The system clock however I assume may be a timer or a crystal that is fed into the bus and peripherals.

pizi-nordic commented:
In the proposed changes I used the following scheme:
- System Clock (referenced through clocks property from the peripherals) is the speed of bus driving peripherals (for example for nRF52xxx this is 16MHz).
- CPU Clock (set through clock-frequency property in CPU nodes) is the speed of the CPU (for example for nRF52xxx this is 64MHz).

All hardware details are hidden in this abstraction. So neither value represents actual crystal.

pizi-nordic committed:
dts: Add information about CPU frequency to the cpu nodes

This commit adds 'clock-frequency' property to the cpu nodes.
The clock frequency specified in the added property is used
during platform configuration. Examples:

- The SWO logger uses clock frequency to configure SWO output.
- Plenty of platforms need CPU clock specified for their HAL.
- Most of devices with USB needs information about CPU clock
  in order to configure USB clock source.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
dts: Add information about system bus frequency to the dts

This commit adds a fixed clock node (representing clock driving
system bus). The added node is then referenced by peripherals requiring
information about driving clock frequency.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

carlescufi commented:
@mnkp the boards you requested in the previous PR are addressed:

```
boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.dts
boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts
dts/arm/silabs/efr32mg.dtsi
dts/arm/silabs/efr32fg1p.dtsi
```

pizi-nordic commented:
I addressed all comments. Silabs, Atmel and ARC targets are updated. Please review.

mnkp reviewed (commented):
Continuing discussion from  #16469 regarding usage of "fixed-clock" DTS nodes:

> As agreed with @galak, my changes are not intended to abstract SoC clock tree. As I do not know all the hardware details, I only created black box(es) representing given clocks. For most of the SoCs, these clocks will have the same frequency as the CPU clock but this is not always the case - this is why bus clock is separated from CPU clock.

Implementing proper clock tree in DTS requires intricate knowledge of underlying hardware. That's certainly out of scope of this PR. However, as I argued
> The "fixed-clock" DTS nodes are used [by Linux DTS] primarily for the clocks generated directly by quartz or fixed by design at some predefined frequency. Thus "fixed-clock". They are not used for system clocks which - almost never - are fixed. 

This PR introduced "fixed-clock" nodes that define system clock not because it's the right thing to do but, as far as I can tell, because it makes it easy to extract properties from DTS and generate required defines. I.e. we shouldn't use "fixed-clock" nodes as black boxes to represent system clocks.

So, my proposal is to do less in this PR. Rather than adding "fixed-clock" DTS nodes everywhere where we need to get rid of `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` defines we can replace it by `DT_CPU_CLOCK_FREQUENCY` define. This would be enough to decouple system clock frequency as seen by the kernel from CPU frequency and the clock tree that depends on it. SoC maintainers can take it from there and implement proper DTS clock tree, if needed.

pizi-nordic commented:
> We shouldn't use "fixed-clock" nodes as black boxes to represent system clocks.
I have tried this approach (https://github.com/zephyrproject-rtos/zephyr/pull/15514, https://github.com/zephyrproject-rtos/zephyr/pull/15505, https://github.com/zephyrproject-rtos/zephyr/pull/15424, https://github.com/zephyrproject-rtos/zephyr/pull/15415, https://github.com/zephyrproject-rtos/zephyr/pull/15356). @galak rejected this.

> So, my proposal is to do less in this PR. Rather than adding "fixed-clock" DTS nodes everywhere where we need to get rid of CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC defines we can replace it by DT_CPU_CLOCK_FREQUENCY define. 

We cannot do that! Peripherals are not clocked form the CPU, so we cannot use the DT_CPU_CLOCK_FREQUENCY: https://github.com/zephyrproject-rtos/zephyr/pull/16957#issuecomment-504392207.

I know, that there are chips which have the save frequency on CPU and Bus, but we cannot have such assumption in the code.

mnkp commented:
> We cannot do that! Peripherals are not clocked form the CPU, so we cannot use the DT_CPU_CLOCK_FREQUENCY

It's already being done. Everywhere where the drivers currently use `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` they use peripheral clock that is derived from the CPU clock.

E.g. that's what `drivers/serial/uart_qmsi.c` does to set up its baud rate.
```
#define DIVISOR_LOW(baudrate) \
	((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / (16 * baudrate)) & 0xFF)
#define DIVISOR_HIGH(baudrate) \
	(((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / (16 * baudrate)) & 0xFF00) >> 8)
```

pizi-nordic commented:
> It's already being done. Everywhere where the drivers currently use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC they use peripheral clock that is derived from the CPU clock.

Not from the CPU clock. The CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is a frequency of system timer. And this was wrong as choosing different kernel clock source or enabling prescaling in the timer would disrupt unrelated things like UART, SPI etc. This is exactly the problem being fixed by this PR.

Here is the example from the ./arc/arc_iot.dtsi. There are 2 UARTs clocked by different source:
```
                uart2: uart@80014200 {
                        compatible = "ns16550";
                        clock-frequency = <16000000>;
                        reg = <0x80014200 0x1000>;
                        label = "UART_2";
                        interrupts = <88 0>;
                        interrupt-parent = <&intc>;

                };

                uart3: uart@80014300 {
                        compatible = "ns16550";
                        clock-frequency = <144000000>;
                        reg = <0x80014300 0x100>;
                        label = "UART_3";
                        interrupts = <89 0>;
                        interrupt-parent = <&intc>;

                };
```

As you see we cannot use CPU clock frequency everywhere just because it is equal to the system bus frequency on your SoC...

carlescufi commented:
@mnkp as Piotr says, `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` is not the CPU frequency, but the system timer frequency. On Nordic for example it's 32768, whereas the CPU typically runs at 64MHz and the system bus at 16MHz

jfischer-phytec-iot commented on dts/bindings/usb/nxp,kinetis-usbd.yaml:
```diff
@@ -22,3 +22,9 @@ properties:
     interrupts:
       category: required
+
+    clocks:
```
@pizi-nordic Should it be part of this commit?

carlescufi replied:
Yes, this commit also adds a couple of .yaml bindings

jfischer-phytec-iot replied:
No it just two, `nxp,kinetis-usbd.yaml` and `arm,cmsdk-watchdog.yaml`, the second is okay (v2m_musca-common.dtsi), but where is the `clocks` from `nxp,kinetis-usbd.yaml` used?

pizi-nordic replied:
Here: https://github.com/zephyrproject-rtos/zephyr/pull/16958/files#diff-be3b6b842e8552442c99fc8abcafeafd

carlescufi replied:
Yes, we had to split up the .dts and .yaml additions and the code changes to simplify the PR

jfischer-phytec-iot replied:
okay I understood the changes here.

galak replied:
This is wrong, and needs a fix elsewhere

ioannisg reviewed (commented):
The Cortex-M23, 33 bindings are not added. Why?

carlescufi commented:
> The Cortex-M23, 33 bindings are not added. Why?

Not sure why, but I will add them.

ioannisg reviewed (commented):
The PR looks ok to me, now.
I've just got some rather generic questions:
- What has been the rationale behind adding the clock-frequency property in the board .dts file, instead of the SOC dts headers? As far as I understand from the comments of @pizi-nordic here (correct me if I am wrong), the clock-frequency is the frequency the CPU runs. So I would expect this is always put in the SOC dts, provided that such dts header exists, of course. Am I missing something?
 - Why the clock-frequency is not added for nRF SoC/Boards?

galak commented on dts/bindings/cpu/cpu.yaml:
```diff
@@ -0,0 +1,27 @@
+      description: Clock frequency in Hz
+      generation: define
+
+    clocks:
```
Is there a case we have a clock property on a cpu node?

carlescufi replied:
No, I checked. Should I remove this for now?

carlescufi replied:
Checked with @pizi-nordic. According to the spec the frequency either comes from `clock-frequency` or from `clocks` so it is here for completeness with the spec.

galak replied:
Ok, leave it in than.

galak reviewed (commented):
a few questions/minor comments.

carlescufi commented:
> * What has been the rationale behind adding the clock-frequency property in the board .dts file, instead of the SOC dts headers? As far as I understand from the comments of @pizi-nordic here (correct me if I am wrong), the clock-frequency is the frequency the CPU runs. So I would expect this is always put in the SOC dts, provided that such dts header exists, of course. Am I missing something?

Because the CPU runs at a frequency that is defined by the board, not the core. The core can run at multiple frequencies, but the board and its crystals ultimately decide the CPU clock itself.

> * Why the clock-frequency is not added for nRF SoC/Boards?

Because it wasn't needed since the nRF code wasn't using the wrong macros, it was already using the correct ones. This is a first pass, we will send more PRs in the near future

carlescufi commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -341,9 +347,10 @@
 			num-bidir-endpoints = <8>;
 			maximum-speed = "full-speed";
-			status = "disabled";
+			status = "ok";
```
Good question. Since @pizi-nordic is not here right now, @jfischer-phytec-iot do you know the reason?

carlescufi replied:
I checked with @pizi-nordic. This is required because with the USB disabled the clock information is not available and it is required during SoC setup

galak replied:
Where is that soc code?  Seems like its missing an ifdef for USB driver being enabled or something.

carlescufi replied:
Here: https://github.com/zephyrproject-rtos/zephyr/pull/16958/files#diff-be3b6b842e8552442c99fc8abcafeafd

MaureenHelm replied:
I don't understand why this is needed. Can you point me to a test or sample that fails sanitycheck without this change? The mimxrt10{15,20,60} boards don't yet enable usb in their board.yaml files, so sanitycheck should not be trying to build `samples/subsys/usb` for them.

pizi-nordic replied:
It failed on mimxrt1050_evk after applying https://github.com/zephyrproject-rtos/zephyr/pull/16958/files#diff-be3b6b842e8552442c99fc8abcafeafd.

But I see that our scripts started to generate clock information for disabled nodes, so the problem no longer exists.

I have removed the change of status from this PR.

erwango commented on dts/bindings/clock/fixed-clock.yaml:
```diff
@@ -0,0 +1,41 @@
+      category: required
+      description: compatible strings
+      constraint: "fixed-clock"
+
```
Don't we miss a "clocks-controller" property here?

galak replied:
It appears that `clocks-controller` is not spec'd in any upstream binding.  Do we need it in Zephyr?  Guessing we use it in the generator.

galak replied:
Hmm, doesn't look like we use it in the generator from a simple grep.

carlescufi replied:
No, I agree, we are not using it. We use `fixed-clock` and then `clock-frequency`

galak commented:
* We need to clarify for SWO in Kconfig or somewhere the use of DT CPU CLOCK FREQ
* The i.mx RT USB DT enablement is wrong.  This is masking some other bug that needs fixing, need to understand what actually fails the 'status' is left as is ('disabled').
* Need to scrub PR for which DTS actually truly need the cpu clock frequency (for example ARC appears not to use it anywhere based on the other PR).

pizi-nordic commented:
> The PR looks ok to me, now.
> I've just got some rather generic questions:
> 
>     * What has been the rationale behind adding the clock-frequency property in the board .dts file, instead of the SOC dts headers? As far as I understand from the comments of @pizi-nordic here (correct me if I am wrong), the clock-frequency is the frequency the CPU runs. So I would expect this is always put in the SOC dts, provided that such dts header exists, of course. Am I missing something?

This was done on @mnkp request as for most of the SoC the board (to be more specific: crystal frequency) determines clock frequency.

>     * Why the clock-frequency is not added for nRF SoC/Boards?

Because we are using values from nRFX not from the DT. So there was no need to add them.

pizi-nordic commented:
> Need to scrub PR for which DTS actually truly need the cpu clock frequency (for example ARC appears not to use it anywhere based on the other PR).

Originally only the affected platforms were modified. The ARC definitions were requested during the review and added later.

pizi-nordic commented:
> The i.mx RT USB DT enablement is wrong. This is masking some other bug that needs fixing, need to understand what actually fails the 'status' is left as is ('disabled').

I can #ifdef the affected part of the code to disable it if USB is disabled in DTS, but someone knowing platform should confirm if such solution is correct.

pizi-nordic commented:
> We need to clarify for SWO in Kconfig or somewhere the use of DT CPU CLOCK FREQ

This will be done in https://github.com/zephyrproject-rtos/zephyr/pull/16958.

pizi-nordic commented:
@galak: I think that all comments are addressed. I have analysed the USB case and there is no change in the behaviour. The same part of SoC code is executed with the same arguments depending only on Kconfig settings. So this PR does not change anything in the area.

galak commented:
> @galak: I think that all comments are addressed. I have analysed the USB case and there is no change in the behaviour. The same part of SoC code is executed with the same arguments depending only on Kconfig settings. So this PR does not change anything in the area.

There is something wrong if you need to change the `status` field for the USB case.  Can you explain why or what causes that change to be required.

galak commented:
In the 'CPU frequency' commit can you:
* Remove `clock-frequency` from the arc soc's as I don't think anything needs it
* Can you explain in the commit message what possible users are there (ie on Xtensa its will be used to define XT_CLOCK_FREQ, on ARM its used for SWO.
* What was the driving factor for setting `clock-frequency` on the given ARM platforms that it get set on?

pizi-nordic commented:
@galak: Old code used system timer frequency as USB controller clock frequency. This is not correct and there is no doubt. So I added the required clock information to USB node and used that value in the code. However with disabled device the define holding clock information was not generated, so there was compilation error. So I changed the status in the DTS.

If this is a real problem for you, I can replace the USB controller clock by the CPU clock in the soc code (both have the same value on that platform), but this PR is intended to solve problems like this...

galak commented:
> @galak: Old code used system timer frequency as USB controller clock frequency. This is not correct and there is no doubt. So I added the required clock information to USB node and used that value in the code. However with disabled device the define holding clock information was not generated, so there was compilation error. So I changed the status in the DTS.
> 
> If this is a real problem for you, I can replace the USB controller clock by the CPU clock in the soc code (both have the same value on that platform), but this PR is intended to solve problems like this...

How can I reproduce the compile error?

galak commented:
> How can I reproduce the compile error?

I think I see the issue, we need @MaureenHelm input on how to handle the enablement of USB on the SoC.  Right now we set `CONFIG_USB_DC_NXP_EHCI` to `y` in `soc/arm/nxp_imx/rt/Kconfig.defconfig.series` regardless if a usb port is enabled in the DTS.  So not sure if it would be better to move that into the board `Kconfig.defconfig`

pizi-nordic commented:
> Right now we set `CONFIG_USB_DC_NXP_EHCI` to `y` in `soc/arm/nxp_imx/rt/Kconfig.defconfig.series` regardless if a usb port is enabled in the DT

I think that every platform has this problem. The Kconfig enables driver and we assume that the driver always find the device. If the device is disabled in DTS, the DT_* defines are missing and we have compilation errors. This is why we keep all out peripheral enabled in the DTS and let Kconfig device what is really used.

pizi-nordic commented:
> Remove clock-frequency from the arc soc's as I don't think anything needs it

Adding ARC was done on request made during review. I would like to skip back and forth updates of the PR...
   
> Can you explain in the commit message what possible users are there (ie on Xtensa its will be used to define XT_CLOCK_FREQ, on ARM its used for SWO.

OK.
 
> What was the driving factor for setting clock-frequency on the given ARM platforms that it get set on?

Decoupling system timer frequency from the parts which depends on system bus / CPU frequency which is accidentally the same as the CPU / Bus frequency. This is preparation for selectable system clock source: https://github.com/zephyrproject-rtos/zephyr/pull/16514

pizi-nordic commented:
>>Can you explain in the commit message what possible users are there (ie on Xtensa its will be used to define XT_CLOCK_FREQ, on ARM its used for SWO.

>OK.

Done.

pizi-nordic commented:
@MaureenHelm: Could you please look at https://github.com/zephyrproject-rtos/zephyr/pull/16957/files#diff-415e26977ba78aa6c06e08282c7ce7d1 and tell if this change is OK for you?

Context is here: https://github.com/zephyrproject-rtos/zephyr/pull/16957#issuecomment-509992728

galak reviewed (commented):
This looks good, but can you update #16958 on top of this to see the changes.

pizi-nordic commented:
> This looks good, but can you update #16958 on top of this to see the changes.

Done. Feel free to look at the changes.

galak commented on dts/bindings/usb/nxp,kinetis-usbd.yaml:
```diff
@@ -22,3 +22,8 @@ properties:
       category: required
+
+    clocks:
+      type: array
```
Now that clocks is in base.yaml, we just need 'category: required' here the type, and description aren't needed.

carlescufi replied:
Done

galak commented on dts/bindings/watchdog/arm,cmsdk-watchdog.yaml:
```diff
@@ -14,3 +14,8 @@ properties:
       category: required
+
+    clocks:
+      type: array
```
Now that clocks is in base.yaml, we just need 'category: required' here the type, and description aren't needed.

carlescufi replied:
Done
